### DISCUSSION TITLE
Linter proposal: warn if returning a typed nil

### DISCUSSION THREAD

**Discussion by mitar** [golangci/golangci-lint#4754] [Linter Ideas]
### Your feature request related to a problem? Please describe.

The problem is well described in [Go FAQ](https://go.dev/doc/faq#nil_error). It is often surprising that comparing a typed interface nil to nil is not true.

### Describe the solution you'd like.

I would like to propose to make a linter which warns if a function is returning a typed nil.

### Describe alternatives you've considered.

There were many potential solutions proposed for Go 2 to address that, but it looks like it is hard to find a backward compatible solution here. But on the other hand, having a typed nil on purpose is a very rare case and something which might be a reasonable false positive for a linter.

**Comment by ldez** [comment]
https://github.com/golang/go/issues/61372

**Comment by mitar** [comment]
@ldez That proposal (at its past and current iteration) does not address the issue this issue is about, because zero can be used only where nil cannot be used. And this issue is about (typed) nil.

**Comment by lixiao189** [comment]
It seems that the linter integrate the Staticcheck now. This module can detect the comparison of interface value with untyped nil.
https://staticcheck.dev/docs/checks/#SA4023
